pulser/simulation/simresults.py

```diff
@@ -108,122 +384,124 @@ def get_final_state(self, reduce_to_basis: Optional[str] = None,
+                weights[0] = 1.
 
         elif self._dim == 3:
-            if meas_basis == 'ground-rydberg':
+            if self._meas_basis == 'ground-rydberg':
                 one_state = 0       # 1 = |r>
                 ex_one = slice(1, 3)
-            elif meas_basis == 'digital':
+            elif self._meas_basis == 'digital':
                 one_state = 2       # 1 = |h>
                 ex_one = slice(0, 2)
-
             probs = probs.reshape([3]*N)
-            weights = []
+            weights = np.zeros(2**N)
             for dec_val in range(2**N):
                 ind: list[Union[int, slice]] = []
                 for v in np.binary_repr(dec_val, width=N):
                     if v == '0':
                         ind.append(ex_one)
                     else:
                         ind.append(one_state)
-                # Eg: 'digital' basis => |1> = index 2, |0> = index 0, 1 = 0:2
+                # Eg: 'digital' basis : |1> = index2, |0> = index0, 1 = 0:2
                 # p_11010 = sum(probs[2, 2, 0:2, 2, 0:2])
-                # We sum all probabilites that correspond to measuring 11010,
-                # namely hhghg, hhrhg, hhghr, hhrhr
-                weights.append(np.sum(probs[tuple(ind)]))
+                # We sum all probabilites that correspond to measuring
+                # 11010, namely hhghg, hhrhg, hhghr, hhrhr
+                weights[dec_val] = np.sum(probs[tuple(ind)])
         else:
             raise NotImplementedError(
                 "Cannot sample system with single-atom state vectors of "
-                "dimension > 3."
-                )
-        dist = np.random.multinomial(N_samples, weights)
-        return {np.binary_repr(i, N): dist[i] for i in np.nonzero(dist)[0]}
+                "dimension > 3.")
+        # Takes care of numerical artefacts in case sum(weights) != 1
+        weights /= sum(weights)
+        return cast(list[float], weights)
+
+    def _detection_from_basis_state(self, n_detections: int, shot: str,
+                                    spam: dict[str, float]) -> Counter:
+        """Computes the distribution of states detected when detecting `shot`.
+
+        Part of the SPAM implementation : computes measurement errors.
+
+        Args:
+            n_detections (int): Number of times state has been detected.
+            shot (str): Binary string of length the number of atoms of the
+            simulation.
+            spam (dict): Dictionnary gathering the SPAM error
+            probabilities.
+        """
+        n_0 = shot.count('0')
+        n_1 = shot.count('1')
+        eps = spam['epsilon']
+        eps_p = spam['epsilon_prime']
+        # Verified
+        prob_1_to_0 = eps_p * (1 - eps) ** n_0 * (1 - eps_p) ** (n_1 - 1)
+        prob_0_to_1 = eps * (1 - eps) ** (n_0 - 1) * (1 - eps_p) ** n_1
+        probs = [int(shot[i]) * prob_1_to_0 + (1 - int(shot[i]))
+                 * prob_0_to_1 for i in range(len(shot))]
```

Comment:
You can actually iterate through the string,
```suggestion
        probs = [prob_1_to_0 if i == '1' else prob_0_to_1 for i in shot)]
```